nsm/scripts/generate-routes.js

```diff
@@ -5,7 +5,144 @@ const prettier = require("prettier")
+  const pagesDir = existsSync(path.resolve(rootDir, "pages"))
+    ? path.resolve(rootDir, "pages")
+    : path.resolve(rootDir, "src/pages")
+  const pageFiles = glob.sync("**/*.{ts,tsx,js,jsx}", { cwd: pagesDir })
+
+  const manifest = {}
+
+  for (const pageFile of pageFiles) {
+    const relativePagePath = path.relative(process.cwd(), pageFile)
+    const parsedPagePath = path.parse(relativePagePath)
+
+    const isInvalidExtension = ![".tsx", ".jsx", ".ts", ".js"].includes(
+      parsedPagePath.ext
+    )
+
+    if (isInvalidExtension) {
+      throw new Error(
+        `Invalid file extension for page: ${pageFile}, must be one of: .tsx, .jsx, .ts, .js. Found: ${pageName.ext}`
+      )
+    }
+
+    const filenameWithoutExtension = parsedPagePath.base
+      .replace(parsedPagePath.ext, "")
+      .replace("index", "")
+
+    const filePathWithoutExtension = path.join(
+      parsedPagePath.dir,
+      filenameWithoutExtension
+    )
+
+    manifest[`/${filePathWithoutExtension}`] = `pages/${relativePagePath}`
+  }
+
+  const manifestJSON = JSON.stringify(manifest, null, 2)
+
+  const manifestFilePath = path.join(outputDir, "pages-manifest.json")
+
+  // there is no need to write the manifest file
+  // but it's good to have it for debugging purposes
+  await fs.writeFile(manifestFilePath, manifestJSON)
+
+  console.log(`Pages manifest generated at: ${manifestFilePath}`)
+
+  return manifest
+}
+
+function generateRouteFile({
+  pagesDirRelativePath,
+  pagesManifest,
+  skipNextBuild,
+  ignoreNotApiFiles,
+  staticFiles,
+}) {
+  const staticFilesString = skipNextBuild
+    ? ""
+    : `,
+  ${staticFiles
+    .filter((fp) => fp.startsWith("static/"))
+    .map(
+      (fp) =>
+        `"/_next/${fp}": serveStatic("${
+          fp.split(".").slice(-1)[0]
+        }", require("./generated_static/${fp}.ts").default)`
+    )}`
+
+  return prettier.format(
+    `// @ts-nocheck
+    import serveStatic from "./serve-static"
+    export default {
+  ${Object.entries(pagesManifest)
+    .filter(([route, fp]) => {
+      if (!skipNextBuild) {
+        return true
+      }
+
+      if (!ignoreNotApiFiles && !fp.startsWith("pages/api")) {
+        throw new Error(
+          `At the moment, only API routes are supported. Found: ${fp}`
+        )
+      }
+
+      if (
+        (ignoreNotApiFiles && !fp.startsWith("pages/api")) ||
+        (!skipNextBuild && ignoreNotApiFiles && !fp.startsWith("pages/api"))
+      ) {
+        return false
+      }
+
+      return true
+    })
+    .map(([route, fp]) => {
+      const fpNoExt = fp.split(".").slice(0, -1).join(".")
+      const fpExt = fpNoExt.split(".").slice(-1)[0]
+
+      return fp.startsWith("pages/api")
+        ? `"${route}": require("${pagesDirRelativePath}/${fpNoExt}")`
+        : `"${route}": serveStatic("${fpExt}", require("./generated_static/server/${fp}.ts").default)`
+    })
+
+    .join(",")}${staticFilesString}
+}
+
+`,
+    { semi: false, parser: "babel" }
```

Comment:
We should respect the consumers prettier config if possible.

Reply:
I personally don't think it's a big deal since the file will be saved in .nsm and the consumers won't be looking at it.